Add a global feature flag to enable the "new" stereo perception code

greglandrum opened the pull request:
This currently defaults to true, but when it is set then the new stereo code is used when calling `AssignStereochemistry()`

greglandrum commented:
@d-b-w this is the PR we discussed yesterday

ricrogz reviewed (approved):
Looks good to me.

ricrogz commented:
> I guess we need to move away from the statically linked python extension modules, right? There's no actual solution to this?

Yes, please. We also had issues with logging due to the static linking, but I guess this is more of a problem.

d-b-w reviewed (commented):
Thanks for this!

I'd like to disable legacy perception globally within the Schrödinger software suite. All tools within the suite are launched from a launcher process that sets up the environment. We have both Python and C++ applications that use RDKit (and many more applications that never use RDKit at all, so I don't want to incur a global load of the RDKit libraries).

For us, it might make sense to have an environment variable to swap the default.

Alternatively, Schrödinger could stack a local commit on top of this that sets the default state of stereo perception.

ptosco commented:
@greglandrum 
> the normal windows Python tests are failing here because the extension modules are statically linked and the "global" I define ends up duplicated in every shared module. Things work fine with windows DLL builds since that doesn't happen there. I guess we need to move away from the statically linked python extension modules, right? There's no actual solution to this?
This would happen on any static build, irrespective of the platform. We can certainly make DLL builds the default also on Windows (that's my personal default anyway :-) ), but we'd probably better disallow static Python builds to avoid that users incur in unexpected behavior if they choose to statically link their Python modules.
The only simple alternative solution I can think of would be to use an environment variable (e.g. `RDK_USE_LEGACY_STEREO_PERCEPTION`), which would encompass the process scope and hence work from both Python and C++ irrespective of the chosen linkage.

greglandrum commented on Code/GraphMol/Wrap/rough_test.py:
```diff
@@ -6891,6 +6891,16 @@ def testDisableNontetrahedralStereo(self):
+    Chem.SetLegacyStereoPerception(False)
+    m = Chem.MolFromSmiles("C[C@H]1CCC2(CC1)CC[C@H](C)C(C)C2")
+    self.assertNotEqual(m.GetAtomWithIdx(1).GetChiralTag(), Chem.ChiralType.CHI_UNSPECIFIED)
+    self.assertNotEqual(m.GetAtomWithIdx(9).GetChiralTag(), Chem.ChiralType.CHI_UNSPECIFIED)
```
yep, will fix both of these things

greglandrum commented:
@ptosco @d-b-w, I'd be happy to swap this (along with the non-tetrahedral stereo flag) to use environment variables instead of globals.

That's easier, doesn't require additional API, and would be thread safe... those are all bonuses.

The downside is that you would no longer be able to infer what a piece of code is going to do by simply looking at it.
I guess we have this to some extent already with parameter files which are loaded using RDBASE, and changing away from the defaults is an advanced-user feature, but it still seems suboptimal.

Thoughts?

greglandrum commented:
> This would allow us to future proof with a configuration singleton in the future without any code change. It would also fix the duplicated global issue in the static and dynamic links I expect.

Unless I'm misunderstanding what you mean, I don't think we'll ever be able to do a configuration singleton which will work in a statically linked environment unless the individual libraries are completely independent of each other (which they aren't and shouldn't be).

bp-kelley commented:
@greglandrum I'm not sure I follow, the Myer's pattern is well defined for both static and dynamic linkage in c++11 and has better features in c++17

```
static Singleton& instance() {
  static Singleton instance;
  return instance;
}
```

This could read from environment variables and the like in a thread safe manner during the construction of the static instance.

d-b-w commented:
I agree that in general it's opaque to use an environment variable for settings that are persistent and that callers will want to change programmatically.

But I think that for these "feature flag" things that expose behaviors that will be default in the future it is useful to expose them as environment variables. That allows an entire process tree to have a consistent state for the feature flag. I can imagine that this might be nice for KNIME users as well as the schrödinger suite use-case I mentioned above. It's also nice future-proofing - the env variable can live past the retirement of its use without causing a program to error.


I think that we should also have programmatic access so that we can demonstrate these in jupyter, for instance.

bp-kelley commented:
Or you could just use the singleton to set the globals from the environment variables if they exist.

```
void _configFromEnvironment() {
  // only initialized once (maybe once per static linkage?)
  static ConfigFromEnvironment instance;
}

bool getUseLegacyStereo() {
  configFromEnvironment();
  return useLegacyStereo;
}

bool getSomeOtherConfig() {
   configFromEngironment();
   return someOtherConfig;
}
```

greglandrum commented:
> @greglandrum I'm not sure I follow, the Meyer's pattern is well defined for both static and dynamic linkage in c++11 and has better features in c++17

@bp-kelley doesn’t that just end up being the same as having a global int in the end? You end up with one instance of the singleton per static linkage unit.

bp-kelley commented:
The static keyword for variables ensures that there is only one per process/program space I  believe.

The exist only in the translation unit they were defined in (i.e. source file in this case)

greglandrum commented:
I've changed the implementation to use environment variables, added getters and setters for both variables for both C++ and Python, and added some documentation to `RDKit_Book.rst` explaining the feature flags.
This version works for me on Linux with dynamic linkage and Windows with static linkage, but let's see what the CI builds do.

d-b-w reviewed (approved):
Thanks, Greg! I'm excited for this!

greglandrum commented on Code/GraphMol/Chirality.cpp:
```diff
@@ -23,6 +23,9 @@
 #include <Geometry/point.h>
 #include "Chirality.h"
 
+#include <cstdlib>
+#include <stdlib.h>
```
ah, right... will fix

bp-kelley commented:
@greglandrum as long as the static is defined in a cpp file ( and not a header file ) I believe it is one per process space.  But that is easy to resolve and not needed for this PR now that we have an abstraction.
